### ISSUE TITLE
sidecar hangs and doesn't process new containers

### ISSUE THREAD

**Issue by nonsense** [ipfs/testground#391]
After running `testground-sidecar` for a while it stops processing new container requests.

Reproduce
---
1. Run `sidecar daemonset` with latest version `ipfs/testground:edge`
2. Run `network/ping-pong` test -> it passes successfully.
3. Try to run `dht/find-peers` test a few times -> currently this fails every time.
4. Try to run `network/ping-pong` test again -> it fails.

Upon inspection of `network/ping-pong` containers and `sidecar` containers, it appears that `network/ping-pong` is blocked at `before WaitNetworkInitialized`:
```
kubectl logs tg-network-6pjbq
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"230f957d-de77-4247-9654-9ae959a5b639","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.39.0.0","Mask":"//8AAA=="}},"timestamp":1579705930021260972,"msg":"before sync.MustWatcherWriter"}
{"runenv":{"test_plan":"network","test_case":"ping-pong","test_run":"230f957d-de77-4247-9654-9ae959a5b639","test_seq":0,"test_instance_count":2,"test_sidecar":true,"test_network":{"IP":"10.39.0.0","Mask":"//8AAA=="}},"timestamp":1579705930026266998,"msg":"before sync.WaitNetworkInitialized"}
```

On the `sidecar` daemon on the given host there are no errors, but it is not processing any requests for IPs on subnet `10.39.0.0`.

Rebooting `sidecar` and trying the test again solves the issue.

**Comment by Stebalien** [comment]
Could we get a stack dump from the sidecar?

**Comment by nonsense** [comment]
Next time this happens, I will attach to the container and get it.

**Comment by nonsense** [comment]
This happened again, after running a few 1k testplans... at some point I noticed that testplan containers don't have the data plane network attached, and searched for one particular container ID in the corresponding sidecar logs, and found nothing - I should have seen `ignoring container` or `got container`.

[sidecar-logs.txt](https://github.com/ipfs/testground/files/4117359/sidecar-logs.txt)

**Comment by Stebalien** [comment]
I'm seeing  a SIGABORT. That usually, in my experience, means an OOM.

**Comment by Stebalien** [comment]
I'm seeing 424 redis connection pool reaper goroutines. That could be holding some memory.

@nonsense could we get a heap dump (pprof) from the sidecar while it's running? We'll need to add a pprof HTTP endpoint to the sidecar to make that work.

**Comment by Stebalien** [comment]
https://github.com/ipfs/testground/issues/423

**Comment by nonsense** [comment]
> I'm seeing a SIGABORT. That usually, in my experience, means an OOM.

That's the signal I sent so that I trigger the stack trace.

**Comment by Stebalien** [comment]
Ah.

---

Errors:

```
Jan 27 13:32:56.455065	[31mERROR[0m	sidecar worker failed: failed to initialise the container: not running	{"host": "unix:///var/run/docker.sock", "container": "764c15e3b5661569e676c3639ceb7352273d24ff4a0e3fd3583792a8e947e37d"}
```

This looks like the docker image is shutting down before the sidecar can complete its work.

```
Jan 27 13:02:01.081794	[31mERROR[0m	sidecar worker failed: failed to initialise the container: inspect failed: error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/containers/044460ece4f60308f13a38acd859b3c70c2a7cd74042c44e43c02c49b9140378/json: context canceled	{"host": "unix:///var/run/docker.sock", "container": "044460ece4f60308f13a38acd859b3c70c2a7cd74042c44e43c02c49b9140378"}
```

Same issue but in this case the sidecar has received the context cancellation first.


```
Jan 27 13:01:38.264151	[31mERROR[0m	sidecar worker failed: container worker failed: failed to wait for network ready: context deadline exceeded waiting on network-initialized; not enough elements, required: 1000, got: 970	{"host": "unix:///var/run/docker.sock", "container": "270e740be1ca7ee8733e5676c742f9ab42e25f6cd3cd0a62071b443af0c2ee98"}
```

Same: This looks like the container shutdown while we were waiting for the network to initialize. Note: "deadline exceeded" isn't correct. We get this error whenever the context aborts, even when it has been canceled and there are no deadlines.

---

I'm noticing a lot of the "failed to initialize the container: not running" errors towards the end of that log but that's probably not related.

**Comment by Stebalien** [comment]
Just making sure: we're not doing any autoscaling, are we?

**Comment by nonsense** [comment]
No, we are not doing autoscaling. Generally when I autoscale a cluster, I then delete the `sidecar`, because it is responsible for creating the route to DNS right now, and if this `initContainer` fails (and currently it does), then we have a problem. the `initContainer` has to be improved.

Nevertheless `sidecar` still hangs often (specially on `master`) and stops processing new containers - maybe it loses connection to docker.sock due to an error - I am investigating now.

**Comment by Stebalien** [comment]
@nonsense would you like to try pair debugging this? That'll be a lot faster than ramping up on k8s as I can't reproduce this issue locally.

**Comment by nonsense** [comment]
@Stebalien I also have a hard time reproducing this since Friday - specially when using `run-cfg keep_service=false`, after switching from `job` to `n pods`. I will keep you posted on progress, and it might be a good idea to pair, once we have a way of reproducing it quickly.

**Comment by Stebalien** [comment]
I wonder if we're panicing then hanging in a defer.

**Comment by Stebalien** [comment]
Ah, nevermind. We're probably not obeying a context somewhere. If we get a "node stop" event, we wait for that node handler to return.

**Comment by Stebalien** [comment]
Guess: sync service. We aren't taking contexts _anywhere_. We should probably just pass in a global context and be done with it.

**Comment by nonsense** [comment]
I have been trying for 2 hours now to reproduce this, and I can't.

My process was the following:
1. run `dht/find-peers` with 60 instances - outcome ok
2. run `dht/find-peers` with 60 instances - outcome ok
3. run `dht/find-peers` with 60 instances - force delete pods when sidecar gets events for them - outcome none, and sidecar reports errors as expected.
4. run `dht/find-peers` with 60 instances - outcome ok
5. resize cluster from 4 to 14 machines
6. run `dht/find-peers` with 1000 instances - outcome `context deadline exceeded waiting on bootstrap-connected; not enough elements, required: 1000, got: 996`
7. run `dht/find-peers` with 60 instances - outcome ok

---

Bottom line, I suggest we close this until we notice the problem again.

**Comment by nonsense** [comment]
This is still an issue on `master` :/ We just saw it while pair programming with @aschmahmann

**Comment by Stebalien** [comment]
Anything more in the logs? The sidecar now logs when it hangs waiting for a worker to stop.

**Comment by nonsense** [comment]
@Stebalien looks like this is happening when we force delete pods. I guess the sidecar doesn't detect that very well right now:

```
Feb 10 15:20:20.161907  ERROR   sidecar worker failed: container worker failed: failed to wait for network ready: context deadline exceeded waiting on network-initialized; not enough elements, required: 40, got: 22  {"host": "unix:///var/
run/docker.sock", "container": "821537de6b01d8af5c9485404553d42e8404cb4fdbb4e4c37f3bc10a4f2a26ea"}
Feb 10 15:20:20.161945  ERROR   sidecar worker failed: container worker failed: failed to wait for network ready: context deadline exceeded waiting on network-initialized; not enough elements, required: 40, got: 22  {"host": "unix:///var/
run/docker.sock", "container": "d32e02f549633646506a6f7c057274e3791cb9ac7dd329ae922130247cb97225"}
Feb 10 15:20:20.162035  ERROR   sidecar worker failed: container worker failed: failed to wait for network ready: context deadline exceeded waiting on network-initialized; not enough elements, required: 40, got: 22  {"host": "unix:///var/
run/docker.sock", "container": "bfe5e1c2a9b43da1040accc4590acb063448527b2b3aa55c80d826cf4bc89b99"}
Feb 10 15:20:20.162119  ERROR   sidecar worker failed: container worker failed: failed to wait for network ready: context deadline exceeded waiting on network-initialized; not enough elements, required: 40, got: 22  {"host": "unix:///var/
run/docker.sock", "container": "85227d8bc400fa5fdd20f29c337c1d474043d9dba1581a78c1b174e5e77ad6e3"}
Feb 10 15:20:20.162290  ERROR   sidecar worker failed: container worker failed: failed to wait for network ready: context deadline exceeded waiting on network-initialized; not enough elements, required: 40, got: 22  {"host": "unix:///var/
run/docker.sock", "container": "f667d392ecaddd972f2d8450f44e7f09d72cdce760c1927aae0ff57ed4677f46"}
Feb 10 15:23:10.637483  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:15.637497  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:20.637494  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:25.637494  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:30.637495  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:35.637489  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:40.637493  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:45.637494  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:50.637494  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:23:55.637490  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
Feb 10 15:24:00.637512  ERROR   waiting for container worker to stop    {"host": "unix:///var/run/docker.sock", "container": "fa757219939d8c7e3f1636080524f4125c7e53f047dd6aae25fe92fc0dccb954"}
```

I will be looking at this this week and will have more insight.

**Comment by Stebalien** [comment]
Not if sure I understand. If we force-delete a container that the sidecar is waiting for, the sidecar's context should be canceled and it should abort. Previously, this _wouldn't_ happen because we hadn't piped through the contexts. Now that we're piping the contexts through, this _should_ happen (I think?).

**Comment by nonsense** [comment]
Aaaaand I am closing this again - I ran a lot of tests and haven't encountered this problem since we fixed the k8s runner during team week.
